providers/docker Add docker_registry

xsellier opened the pull request:
To fix this issue: https://github.com/hashicorp/terraform/issues/4934

Commit message:
- Add docker_registry resource
- Add tests for docker_registry
- Add documentation for docker_registry
- Add link between docker_image and docker_registry
- Add test for docker_image
- Add documentation for docker_image
- Fix issue with docker_image (repository parsing)
- TODO: Manage more than one manual docker_registry configuration

xsellier commented:
- Set docker_registry as 'ForceNew' in order to read the file at each deployment (token expired)
- fix docker_registry documentation
- Replace 'auth' by 'auths' in order to be able to manage more than one authentication per docker_registry resource

xsellier commented:
Tested many times during this week, this Pull Request is finally ready (Sorry, I forget the [WIP] tag).
Everything seems to work fine.

apparentlymart commented on website/source/docs/providers/docker/r/registry.html.mdown:
```diff
@@ -0,0 +1,55 @@
+page_title: "Docker: docker_registry"
+sidebar_current: "docs-docker-resource-registry"
+description: |-
+  Logs to a Docker registry, downloads and exports a Docker image.
```
Is this description accurate? It seems like maybe this is the description from `docker_image`.

apparentlymart commented on website/source/docs/providers/docker/r/registry.html.mdown:
```diff
@@ -0,0 +1,55 @@
+<a id="auths"></a>
+### Auths
+
+`auths` is a block within the configuration that can be repeated to specify
```
If each of these blocks represents "one auth" then the usual style is to name it in the singular so that each block appears with a singular heading in the config:

```hcl
auth {
    server_address = "..."
    username = "..."
    password = "..."
}
auth {
    server_address = "..."
    username = "..."
    password = "..."
}
```

I'd agree that it feels a bit weird in the implementation to have a list/set field with a singular name, but it makes sense given how this gets written in the configuration.

xsellier commented:
Any news on this ?

xsellier commented:
Sorry for the double post, but is there any news on this pull request ?

matthewvalimaki commented on website/source/docs/providers/docker/r/image.html.markdown:
````diff
@@ -16,9 +16,19 @@ them.
 ```
+# Log to a custom docker registry
+resource "docker_registry" "ubuntu_private" {
+    auths = {
````
This should be `auth`. Currently it gives `invalid or unknown key: auths`.

xsellier commented:
Nice catch !
I've amended the commit with this fix.

matthewvalimaki commented:
I'm getting `Unable to create container: API error (500): No command specified` when I'm trying it with 
```
resource "docker_image" "registry" {
    name = "alpine:3.1"
    registry = "${docker_registry.registry.configurations}"

}
```
Also trying to use my own hub.docker.com private repository leads to `Unable to read Docker image into resource: Unable to pull image matthewvalimaki/repo:latest`. I do not spot any mistakes in my configuration.

Tested with Docker version 1.10.3.

xsellier commented:
I've tested with docker:
`Docker version 1.11.1, build 5604cbe`

xsellier commented:
Issue is not from docker itself, because it work with 1.10.3.
So far I've tested it without specifying an image tag, tomorrow I will to fix that issue.

Good catch @matthewvalimaki

xsellier commented:
Hey @matthewvalimaki ,

I think I've fixed the issue. It was about the image string parsing. I've replaced it with a regular expression, it should work like a charm now.

Hope you can test it by the end of the week.

matthewvalimaki commented:
@xsellier it does seem to work but not as I expected.
I have to use `server_address = "https://index.docker.io/v1/"` for it to work with Docker Hub. If I try what is in readme `https://hub.docker.com/` it gives `Error pulling image, registry: 'index.docker.io/v1/', repository:`.

xsellier commented:
@matthewvalimaki Oh sure, I'll fix that also tomorrow :)

xsellier commented:
@matthewvalimaki now it should work properly as expected.

Since we can read the credentials from a file, we can have more than one credential loaded. That being said I guess you understand we have to match the `docker_registry` with the `docker_image`. So I added an exception, if there is only one credential loaded from the file, we will use it anyway.

xsellier commented:
any news on this ?

alexraju commented:
This is a much awaited essential. @xsellier Thank you.

lpmaurice commented:
+1

xsellier commented:
Hey folks,

this patch works pretty well, anything new about it ?

radeksimko commented:
Hi @xsellier 
thank you for taking the time to submit this PR.

Although I understand that the only reason user usually authenticates in Docker is to pull images from private docker registry, authentication & connection configuration is usually handled in `provider` rather than `resource` in Terraform.

It may also be helpful for you if you read my comment about the separation of Docker data sources/resources [here](https://github.com/hashicorp/terraform/pull/7000#issuecomment-231596846).
Admittedly it cannot be _completely_ separated due the fact that to pull image from private registry via Docker Remote API one still has to authenticate against the Docker Registry API - but in reality the Docker Remote API only requires the base64 encoded auth string, so technically we're not calling the registry API directly.

https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/#/create-an-image

![screen shot 2016-07-10 at 17 31 24](https://cloud.githubusercontent.com/assets/287584/16714851/70e239a0-46c7-11e6-98b3-ed4108dac61f.png)


I think that the problem described in https://github.com/hashicorp/terraform/issues/4934 is caused by the fact that we don't make authentication explicit -> that is basically what you're fixing here as part of this PR 😃  - except that the logic is inside a new resource rather than part of existing provider.

What do you think about the following?

```hcl
provider "docker" {
  host = "tcp://127.0.0.1:1234/"
  registry_auth = {
    username = "foo"
    password = "bar"
    server_address = "docker.io"
  }
}
```

xsellier commented:
Thank you for this answer @radeksimko .

Although this seems to be a good idea, I think we miss a point.
I use a registry independantly of the provider, my provider allows me to connect to a docker instance and to deploy a docker container. And, I use a `docker_registry` to pull an image Because I want to be able to use many `docker_registry` for a unique `docker_provider` I think we need a `docker_registry` tag.

Let me explain, in my specific case, I use a `docker_provider` to deploy containers. But some of the containers are on different private `docker_registry`. Because I have a _common_ registry, for common containers, and _specific_ registries for specific containers, but everything is deployed on the very same machine (the same `docker_provider`)

Although we can set more than one `registry_auth` per provider, but I also want to factorize the code, so I have a unique terraform file with `docker_registry` credentials, and I use it into others terraform files. And because I don't want to duplicate any `docker_repository`, I though it was a smart move to define a `docker_registry`.

radeksimko commented:
@xsellier I see, thank you for the explanation. Would defining multiple `provider`s help in such cases?

```hcl
variable "docker_host" {
  default = "/my/lovely/docker/unix/socket"
}

provider "docker_registry" {
  alias = "dockerhub"
  host = "https://registry.hub.docker.com"
  username = ""
  password = ""
}

provider "docker_registry" {
  alias = "quay"
  host = "https://quay.io"
  username = ""
  password = ""
}

data "docker_registry_image" "nomad" {
  provider = "docker_registry.dockerhub"
  name = "nomad"
  tag = "latest"
}

data "docker_registry_image" "nomad" {
  provider = "docker_registry.quay"
  name = "coreos"
  tag = "latest"
}

provider "docker" {
  alias = "dockerhub"
  host = "${var.docker_host}"
  auth {
    host = "https://registry.hub.docker.com"
    username = ""
    password = ""
  }
}

provider "docker" {
  alias = "quay"
  host = "${var.docker_host}"
  auth {
    host = "https://quay.io"
    username = ""
    password = ""
  }
}

# First use case
resource "docker_image" "nomad" {
  provider = "docker.dockerhub"
  name = "${data.docker_registry_image.nomad.name}:${data.docker_registry_image.nomad.digest}"
}

# Second use case
resource "docker_image" "nomad" {
  provider = "docker.dockerhub"
  name = "my_special_app:latest"
  trigger_pull = ["${data.docker_registry_image.nomad}"]
}

# Third use case
resource "aws_ecs_task_definition" "nomad" {
  provider = "docker.dockerhub"
  family = "nomad"
  container_definitions = <<DEFINITION
[
  {
    "name": "jenkins",
    "image": "${data.docker_registry_image.nomad.name}:${data.docker_registry_image.nomad.digest}",
    "cpu": 10,
    "memory": 500,
  }
]
DEFINITION
}
```

I wouldn't mind keeping the data source under existing `docker` provider after all, but I still believe that auth should be handled as part of the provider, not as a separate resource. Also if we want the data source to be useful w/out direct docker connection I think we would need to make `host` field in `docker` provider Optional. There are people using container schedulers and these don't talk over Docker Remote API.

Resources have CRUD, hence they Create, Read, Update and Delete things, they typically don't provide authentication - the lifecycle would also prevent us from reusing such _resource_ as authentication source for the data source.

Maybe we should have 1st class support for secrets/credentials, that would be AOK 👌 

```hcl
auth "docker_registry" "hub" {
  host = ""
  user = ""
  pass = ""
}

resource "docker_image" {
  auth = "${auth.docker_registry.hub}"
  name = ""
}
```

Obviously that would require more significant changes in Terraform's core, but it would IMO make sense to support it, especially once encryption of tfstate + Vault integration becomes a reality.

kyhavlov commented:
Potentially dumb question - is there any use in passing the registry creds through the provider directly, like this? For my own use, I put the registry creds in ~/.docker/config.json and use the patch [here](https://github.com/hashicorp/terraform/pull/6892) to load them. The docker config.json already supports creds for multiple registries, and `NewAuthConfigurationsFromDockerCfg()` is already built into the go-dockerclient. I thought letting secrets end up in the terraform state was an anti-pattern, and I do the same thing for the AWS provider (having it pick up the creds via a file instead of passing it into the provider block).

xsellier commented:
@radeksimko I like this syntax, I will implement it soon. Need to re-write some part of the code for that.

Thank you for this review.

radeksimko commented:
> Potentially dumb question - is there any use in passing the registry creds through the provider directly, like this?

I personally wouldn't hard-code username+password, especially because tfstate is currently stored in plaintext, but I can understand why a user would want to specify custom location of `~/.docker/config` and/or custom registry hostname. I'm not saying "let's implement this now" - I'm happy to accept a PR that **doesn't** support `username+password` based auth, but I would not prevent anyone from implementing it & merging it.

Admittedly the registry hostname could be part of `docker_image.name`... which makes me think that `name` may not be the right name for this field... but I'm really nitpicking now... 😄

radeksimko commented:
@xsellier In case it wasn't clear from my example above - the `provider` alias functionality is part of Terraform's core - i.e. should require no extra provider/resource/data-source logic. Just make sure that connections/settings can be shared correctly between a provider and resource/data-source via `meta`.

xsellier commented:
@radeksimko I got you, but I have already spent much time on this feature since few month. It has already been reviewed, so I won't spend more time on this right now, and since I don't know when it's gonna be merged after this review. So I would appreciate it very much if you make the change yourself to be compliant with the very new naming-convention.
